### DISCUSSION TITLE
How to share field definitions between inputObjectType and objectType definitions?

### DISCUSSION THREAD

**Discussion by carnesen** [graphql-nexus/nexus#772] [Q&A] [answered]
For example, suppose I'm trying to implement the following as a Nexus schema:
```
input MessageInput {
  content: String
  author: String
}
 
type Message {
  id: ID!
  content: String
  author: String
}
```
Right now I'm defining the common fields `content` and `author` both explicitly in the `inputObjectType` and `objectType` `definition` blocks. Is there a way in Nexus to pull out the common field definitions?

Thanks in advance,
Chris

**Comment by shreyas44** [comment, marked as answer]
Hey @carnesen! I don't think there's a built in way to achieve what you're talking about. However, you could try using the below workaround. I haven't tested it, so I'm not sure if it would work or not.

```ts
const messageDefinition = (t) => {
  t.string("content")
  t.string("author")
}

const MessageInput = inputObjectType({
  name: "MessageInput",
  defintion: messageDefinition,
})

const Message = objectType({
  name: "Message",
  defintion(t) {
    t.id("id")
    messageDefinition(t)
  }
})
```

**Reply by lmamea** [reply]
This works nicely, thanks!

**Reply by carnesen** [reply]
Thank you for the reply! Is there a TypeScript type we can use for `t` in `messageDefinition`?

**Reply by NEO97online** [reply]
It looks like this is pretty difficult to type without just using `t: any`. I wonder if we could make this easier to type?

**Reply by rozsival** [reply]
@auderer @carnesen You can make use of generics for that case:

```typescript
import {
  InputDefinitionBlock,
  OutputDefinitionBlock,
} from 'nexus/dist/definitions/definitionBlocks'

const messageDefinition = <N extends string>(t: InputDefinitionBlock<N> | OutputDefinitionBlock<N>) => {
  t.string('content')
  t.string('author')
}
```

It's not an ideal solution but it works in most cases. I wish there was some kind of shared interface which defines methods that are usable on both inputs and outputs.

**Reply by RubenMoraAI** [reply]
@rozsival  very good idea, maybe can improve defining a generic type :
`type DefinitionBlock<N extends string> = InputDefinitionBlock<N> | OutputDefinitionBlock<N>;`
`export const wordDefinition = <N extends string>(t: DefinitionBlock<N>) => { ... `

**Reply by Frosty21** [reply]
how could you extend it to a `field` in generics?
